neo4j/src/main/java/org/apache/metamodel/neo4j/Neo4jDataContext.java

```diff
@@ -0,0 +1,336 @@
+        }
+    }
+
+    @Override
+    protected DataSet materializeMainSchemaTable(Table table, Column[] columns, int firstRow, int maxRows) {
+        if ((columns != null) && (columns.length > 0)) {
+            Neo4jDataSet dataSet = null;
+            try {
+                String selectQuery = Neo4jCypherQueryBuilder.buildSelectQuery(table, columns, firstRow, maxRows);
+                String responseJSONString = _requestWrapper.executeCypherQuery(selectQuery);
+                JSONObject resultJSONObject = new JSONObject(responseJSONString);
+                final SelectItem[] selectItems = MetaModelHelper.createSelectItems(columns);
+                dataSet = new Neo4jDataSet(selectItems, resultJSONObject);
+            } catch (JSONException e) {
+                logger.error("Error occured in parsing JSON while materializing the schema: ", e);
+                throw new IllegalStateException(e);
+            }
+
+            return dataSet;
+        } else {
+            logger.error("Encountered null or empty columns array for materializing main schema table.");
+            throw new IllegalArgumentException("Columns cannot be null or empty array");
+        }
+    }
+
+    @Override
+    protected DataSet materializeMainSchemaTable(Table table, Column[] columns, int maxRows) {
+        return materializeMainSchemaTable(table, columns, 1, maxRows);
+    }
+
+    @Override
+    protected org.apache.metamodel.data.Row executePrimaryKeyLookupQuery(Table table, List<SelectItem> selectItems,
+            Column primaryKeyColumn, Object keyValue) {
+        return null;
+    }
+
+    @Override
+    protected Number executeCountQuery(Table table, List<FilterItem> whereItems, boolean functionApproximationAllowed) {
+        String countQuery = Neo4jCypherQueryBuilder.buildCountQuery(table.getName(), whereItems);
+        String jsonResponse = _requestWrapper.executeCypherQuery(countQuery);
+
+        JSONObject jsonResponseObject;
+        try {
+            jsonResponseObject = new JSONObject(jsonResponse);
+            JSONArray resultsJSONArray = jsonResponseObject.getJSONArray("results");
+            JSONObject resultJSONObject = (JSONObject) resultsJSONArray.get(0);
+            JSONArray dataJSONArray = resultJSONObject.getJSONArray("data");
+            JSONObject rowJSONObject = (JSONObject) dataJSONArray.get(0);
+            JSONArray valueJSONArray = rowJSONObject.getJSONArray("row");
+            Number value = (Number) valueJSONArray.get(0);
+            return value;
+        } catch (JSONException e) {
+            logger.error("Error occured in parsing JSON response: ", e);
+            // Do not throw an exception here. Returning null here will make
+            // MetaModel attempt to count records manually and therefore recover
+            // from the error.
+            return null;
+        }
+    }
+
+    @Override
+    public void executeUpdate(UpdateScript script) {
+        throw new UnsupportedOperationException(
```

Comment:
I would just not extend the UpdateableDataContext interface then. This method just gives the wrong impression.